### ISSUE TITLE
[iOS][Simulator]: Trying to toggle software keyboard throws unhandled exception on latest master and beta

### ISSUE THREAD

**Issue by jlubeck** [flutter/flutter#89378]
## Steps to Reproduce

Please check the following screencap (there is no code assigned, it's just toggling the keyboard from the simulator menu.
![keyboard_bug](https://user-images.githubusercontent.com/3067603/131874207-fff9cba3-3c0a-4c99-9cfb-456be361dece.gif)

This is the error in text form:
```
[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 790 pos 16: '(data.physical == 0 && data.logical == 0) ||
               (data.physical != 0 && data.logical != 0)': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      KeyEventManager.handleKeyData (package:flutter/src/services/hardware_keyboard.dart:790:16)
#3      PlatformDispatcher._dispatchKeyData.<anonymous closure> (dart:ui/platform_dispatcher.dart:375:61)
#4      _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#5      _rootRun (dart:async/zone.dart:1428:13)
#6      _CustomZone.run (dart:async/zone.dart:1328:19)
#7      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#8      _invoke2 (dart:ui/hooks.dart:204:10)
#9      PlatformDispatcher._dispatchKeyData (dart:ui/platform_dispatcher.dart:373:5)
#10     _dispatchKeyData (dart:ui/h<…>
[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 790 pos 16: '(data.physical == 0 && data.logical == 0) ||
               (data.physical != 0 && data.logical != 0)': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      KeyEventManager.handleKeyData (package:flutter/src/services/hardware_keyboard.dart:790:16)
#3      PlatformDispatcher._dispatchKeyData.<anonymous closure> (dart:ui/platform_dispatcher.dart:375:61)
#4      _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#5      _rootRun (dart:async/zone.dart:1428:13)
#6      _CustomZone.run (dart:async/zone.dart:1328:19)
#7      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#8      _invoke2 (dart:ui/hooks.dart:204:10)
#9      PlatformDispatcher._dispatchKeyData (dart:ui/platform_dispatcher.dart:373:5)
#10     _dispatchKeyData (dart:ui/h<…>

```

On another note, the only reason I tried using the software keyboard in the first place, is because my own keyboard stopped sending keystrokes into the simulator. So now I can't move past a form unless I copy paste text from outside the simulator. Super annoying... 

<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel master, 2.6.0-1.0.pre.112, on macOS 11.5.2 20G95 darwin-x64, locale en-US)
    • Flutter version 2.6.0-1.0.pre.112 at /Users/jan/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0c65de1616 (3 days ago), 2021-08-30 05:26:01 -0400
    • Engine revision c05688f4bc
    • Dart version 2.15.0 (build 2.15.0-65.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/jan/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 12.5.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.59.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • iPad Pro (9.7-inch) (mobile) • 25E7C8D3-09DE-4A13-84FB-51CEAEBA93C4 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)              • macos                                • darwin-x64     • macOS 11.5.2
      20G95 darwin-x64
    • Chrome (web)                 • chrome                               • web-javascript • Google Chrome
      92.0.4515.159
    ! Error: Errors were encountered while preparing your device for development. Please check the Devices and
      Simulators Window.  (code -1)

! Doctor found issues in 1 category.
```

</details>

**Comment by darshankawar** [comment]
Verified this on latest master and stable and confirmed that I see the same exception working with an iPad only on latest master, because with iphone simulator, I don't see the option, as below:

<img width="561" alt="Screenshot 2021-09-03 at 12 20 46 PM" src="https://user-images.githubusercontent.com/67046386/131964582-7a9b02ca-b1ca-44a4-98d9-9c0ebd37756c.png">


On iPad simulator, trying to toggle keyboard throws below exception:

<details>
<summary>xcode console error log</summary>



```
2021-09-03 12:21:05.412586+0530 Runner[6438:111928] Metal API Validation Enabled
2021-09-03 12:21:05.537472+0530 Runner[6438:112149] flutter: Observatory listening on http://127.0.0.1:51102/GjZSudZe1OI=/
2021-09-03 12:21:12.652650+0530 Runner[6438:112133] [VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 790 pos 16: '(data.physical == 0 && data.logical == 0) ||
               (data.physical != 0 && data.logical != 0)': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      KeyEventManager.handleKeyData (package:flutter/src/services/hardware_keyboard.dart:790:16)
#3      PlatformDispatcher._dispatchKeyData.<anonymous closure> (dart:ui/platform_dispatcher.dart:375:61)
#4      _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#5      _rootRun (dart:async/zone.dart:1428:13)
#6      _CustomZone.run (dart:async/zone.dart:1328:19)
#7      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#8      _invoke2 (dart:ui/hooks.dart:204:10)
#9      PlatformDispatcher._dispatchKeyData (dart:ui/platform_dispatcher.dart:373:5)
#10     _dispatchKeyData (dart:ui/hooks.dart:104:31)
2021-09-03 12:21:12.672162+0530 Runner[6438:112133] [VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 790 pos 16: '(data.physical == 0 && data.logical == 0) ||
               (data.physical != 0 && data.logical != 0)': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      KeyEventManager.handleKeyData (package:flutter/src/services/hardware_keyboard.dart:790:16)
#3      PlatformDispatcher._dispatchKeyData.<anonymous closure> (dart:ui/platform_dispatcher.dart:375:61)
#4      _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#5      _rootRun (dart:async/zone.dart:1428:13)
#6      _CustomZone.run (dart:async/zone.dart:1328:19)
#7      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#8      _invoke2 (dart:ui/hooks.dart:204:10)
#9      PlatformDispatcher._dispatchKeyData (dart:ui/platform_dispatcher.dart:373:5)
#10     _dispatchKeyData (dart:ui/hooks.dart:104:31)

```


</details>

Although on latest stable, the same steps works properly without throwing the exception, the keyboard is toggled properly, as below:


https://user-images.githubusercontent.com/67046386/131964738-495863d4-a684-436f-9eaf-dac050cc83d7.mov


<details>
<summary> stable, master flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 2.2.3, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.2.3 at /Users/dhs/documents/fluttersdk/flutter
    • Framework revision f4abaa0735 (4 days ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • iPhone 12 Pro Max (mobile)                     •
      A5473606-0213-4FD8-BA16-553433949729 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • iPad Pro (12.9-inch) (4th generation) (mobile) •
      AD9F1B85-B622-485E-97E8-E4AC866D5770 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)                                • macos
      • darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)                                   • chrome
      • web-javascript • Google Chrome 92.0.4515.159


• No issues found!



• No issues found!

[✓] Flutter (Channel master, 2.6.0-1.0.pre.168, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.6.0-1.0.pre.168 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision de3e218e48 (6 hours ago), 2021-09-02 18:51:02 -0400
    • Engine revision c02b815416
    • Dart version 2.15.0 (build 2.15.0-69.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • iPhone 12 Pro Max (mobile)                     •
      A5473606-0213-4FD8-BA16-553433949729 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • iPad Pro (12.9-inch) (4th generation) (mobile) •
      AD9F1B85-B622-485E-97E8-E4AC866D5770 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)                                • macos
      • darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)                                   • chrome
      • web-javascript • Google Chrome 92.0.4515.159


• No issues found!




```

</details>

**Comment by stuartmorgan** [comment]
@gspencergoog @dkwingsmt Is this a regression from key handling rewrites?

**Comment by dkwingsmt** [comment]
Likely so. It seems that the iOS embedder somehow converted some event keys into physical 0 or logical 0. I can take a look, but @gspencergoog might have an idea?

**Comment by gspencergoog** [comment]
I don't, we should investigate what events the simulator is sending when it toggles the keyboard.

**Comment by blaneyneil** [comment]
@gspencergoog If it helps to diagnose, there's also a weird but possibly related problem, in 2.5 Stable, with iOS simulator physical keyboards where characters will unexpectedly repeat.  It tends to happen on the 3rd character - typing "who" will result in "whooo" nearly every time.  This doesn't happen on Android simulators nor on iOS devices.    And had never happened to me prior to upgrading to 2.5.0 from 2.2.3.

**Comment by jlubeck** [comment]
> @gspencergoog If it helps to diagnose, there's also a weird but possibly related problem, in 2.5 Stable, with iOS simulator physical keyboards where characters will unexpectedly repeat. It tends to happen on the 3rd character - typing "who" will result in "whooo" nearly every time. This doesn't happen on Android simulators nor on iOS devices. And had never happened to me prior to upgrading to 2.5.0 from 2.2.3.

Yup, there's another issue (also opened by me 😅) about that here: https://github.com/flutter/flutter/issues/86565

**Comment by GitHelge** [comment]
I have the same issue on the iPhone 8 and iPhone 12 simulator when "Toggle Software Keyboard" using "cmd + K" (or with the menu) is pressed. This can only occur if "Connect Hardware Keyboard" is activated previously.

**Comment by milanobrenovic** [comment]
Any update on this? I'm struggling to test my app on the simulator.

**Comment by jonataslaw** [comment]
I think there is a regression that should be done here. It's difficult to test the behavior of textfields in languages other than the keyboard language you're programming in because of this, and it's only resolved by completely closing the app, opening a keyboard in a native app, and recompiling the project. This is exhausting.

**Comment by mcp31** [comment]
`[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 790 pos 16: '(data.physical == 0 && data.logical == 0) ||
               (data.physical != 0 && data.logical != 0)': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      KeyEventManager.handleKeyData (package:flutter/src/services/hardware_keyboard.dart:790:16)
#3      PlatformDispatcher._dispatchKeyData.<anonymous closure> (dart:ui/platform_dispatcher.dart:375:61)
#4      _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#5      _rootRun (dart:async/zone.dart:1428:13)
#6      _CustomZone.run (dart:async/zone.dart:1328:19)
#7      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#8      _invoke2 (dart:ui/hooks.dart:204:10)
#9      PlatformDispatcher._dispatchKeyData (dart:ui/platform_dispatcher.dart:373:5)
#10     _dispatchKeyData (dart:ui/h<…>`

This is the error I got by trying to toggle the keyboard in the iOS simulator. It was working just fine yesterday

**Comment by cycloss** [comment]
I'm getting a complete freeze of my app when tapping a text field using the simulator on iOS 15.0 and Flutter 2.5.1. Was working fine, but i'm not sure whether it was caused by an iOS update, or a Flutter upgrade.

**Comment by bct-fbo** [comment]
Found a temp/easy workaround. You can toggle the software keyboard by clicking on the language icon and select 'Show Keyboard'.

![software_keyboard_toggle](https://user-images.githubusercontent.com/79577203/135422290-7f873a62-feae-423f-ae6d-86a528bf8fa3.gif)

**Comment by SonXuanNguyen2k** [comment]
I encounter the same issue, the iPhone simulator fails to kick in the on-screen keyboard. @darshankawar  to see the "Toggle Software Keyboard" option, u must first enable the "Connect Hardware Keyboard" option with `Shift + Cmd + K` (as shown in my image below).

![Keyboard setting](https://user-images.githubusercontent.com/29955043/135449498-5925c2eb-d8dd-4fe4-b0d2-0125578aaba3.png)
Above is my current keyboard settings for the iPhone simulator. And everytime I try to toggle the on-screen keyboard with `Cmd+K`, an exception is thrown, which prevents the iPhone simulator from showing the keyboard. Here is the full details of the exception:

![Exception details](https://user-images.githubusercontent.com/29955043/135450808-6eb25693-7e98-42c3-a48d-3e7cdfe94efb.png)

The current workaround to make the keyboard pop up is to temporarily disable the "Connect Hardware Keyboard" option, but your simulator will completely ignore all key codes coming from your laptop's keyboard (key codes from simulator's on-screen keyboard are still accepted).

**Comment by lailai0715** [comment]
Same issue here on M1 Mac with arm based iOS simulator

**Comment by pannavanno** [comment]
Same issue with ios simulator.
I am using physical keyboard on my mac.
My current workaround: uncheck "Connect hardware keyboard." then the soft keyboard will be shown.

I recall no need to do this when i develop native app and older flutter version.

Otherwise it will show unhandled exception
````
[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 790 pos 16: '(data.physical == 0 && data.logical == 0) ||
````

**Comment by pro100svitlo** [comment]
facing the same problem!

**Comment by befirst** [comment]
Can't reproduce it with flutter 2.2.3, but can with 2.5.2

**Comment by jmagman** [comment]
```
KeyData(key down, physical: 0x0, logical: 0x13000000b8 (iOS), character: <none>)
```
So physical is 0 and logical is not. #87152
https://github.com/flutter/flutter/blob/4771605c391666f0c7123f85414a6d664a97a0e1/packages/flutter/lib/src/services/hardware_keyboard.dart#L790-L795

@dkwingsmt can you advise what the right thing to do here?  I'm not following why it's important that both be 0 if one is for empty key data.

**Comment by LongCatIsLooong** [comment]
I'm going to take a look. It seems `KeyData.physical` isn't properly populated:

<img width="355" alt="Screen Shot 2021-10-07 at 10 40 00 AM" src="https://user-images.githubusercontent.com/31859944/136435995-130b6a16-e660-40b4-9edf-b214d0d77137.png">

**Comment by Trulsmatias** [comment]
The title of the issue states that this only an issue for the iPad simulator. This is wrong. It is also an issue for the iPhone simulators. Please change 😊

**Comment by jlubeck** [comment]
> The title of the issue states that this only an issue for the iPad simulator. This is wrong. It is also an issue for the iPhone simulators. Please change 😊

Done! I was only testing iPads, so didn't want to make assumptions ;)

**Comment by jmagman** [comment]
> So, still no fix?

Just 👍 the issue, please don't add noise. We're investigating.

**Comment by LongCatIsLooong** [comment]
It seems "Toggle Software Keyboard" simply fires a key down event with the keycode `0xb8` (the eject key on some keyboards) and a corresponding key up event immediately after. To handle the key down event the engine starts a new run loop to block the main thread (except for channel message handlers). But the key up event handling always ends up being scheduled on the inner run loop, effectively making key event processing non-serial.

The pull request linked above should fix the assert, it doesn't fix the "Toggle Software Keyboard" functionality on iOS simulator. This issue should be fixed (or easier to fix) once https://github.com/flutter/engine/pull/29113 lands. **Potential workaround**: you can use "Connect hardware keyboard" to toggle the software keyboard for the time being.

**Comment by vinukn335** [comment]
Remove tick from Keyboard > Use the Same keyboard Language as macOS and Connect Hardware Keyboard.

Flutter 2.5.3, IOS 15.0

![Screen Shot 2021-10-16 at 9 59 54 PM](https://user-images.githubusercontent.com/47185599/137595178-21526b2d-6e3c-48f3-8220-eb321d8de544.png)

**Comment by dkwingsmt** [comment]
@jmagman I don't quite get what is "just prevents the asserts", but yes that PR does close this issue.

**Comment by LongCatIsLooong** [comment]
Sorry didn't mean to close it.

**Comment by albert0m** [comment]
still happening,  Flutter 2.5.3, iOS 15

**Comment by jmagman** [comment]
> still happening, Flutter 2.5.3, iOS 15

The fix is currently available in the master channel.
